scripts/mk_all.lean

```diff
@@ -0,0 +1,108 @@
+      let allLean ← IO.Process.run { cmd := "git", args := #["ls-files", mlDir ++ "*.lean"] }
+      return (((allLean.dropRightWhile (· == '\n')).splitOn "\n").map (⟨·⟩)).toArray
+    else do
+      let all ← walkDir ml
+      return all.filter (·.extension == some "lean"))
+  let files := (allModules.erase ml.lean).qsort (·.toString < ·.toString)
+  let withImport ← files.mapM fun f => do
+    -- this check is helpful in case the `git` option is on and a local file has been removed
+    if ← pathExists f then
+      return "import " ++ (← moduleNameOfFileName f none).toString
+    else return ""
+  return ("\n".intercalate (withImport.filter (· != "")).toList).push '\n'
+
+open Lake in
+/-- `getLeanLibs` returns the names (as an `Array` of `String`s) of all the libraries
+on which the current project depends.
+If the current project is `mathlib`, then it excludes the libraries `Cache` and `LongestPole` and
+it includes `Mathlib/Tactic`. -/
+def getLeanLibs : IO (Array String) := do
+  let (elanInstall?, leanInstall?, lakeInstall?) ← findInstall?
+  let config ← MonadError.runEIO <| mkLoadConfig { elanInstall?, leanInstall?, lakeInstall? }
+  let ws ← MonadError.runEIO (MainM.runLogIO (loadWorkspace config)).toEIO
+  let package := ws.root
+  let libs := (package.leanLibs.map (·.name)).map (·.toString)
+  return if package.name == `mathlib then
+    libs.erase "Cache" |>.erase "LongestPole" |>.push ("Mathlib".push pathSeparator ++ "Tactic")
+  else
+    libs
+
+open IO.FS IO.Process Name Cli in
+/-- Implementation of the `mk_all` command line program.
+The exit code is the number of files that the command updates/creates. -/
+def mkAllCLI (args : Parsed) : IO UInt32 := do
+  -- Check whether the `--git` flag was set
+  let git := (args.flag? "git").isSome
+  -- Check whether the `--lib` flag was set. If so, build the file corresponding to the library
+  -- passed to `--lib`. Else build all the libraries of the package.
+  -- If the package is `mathlib`, then it removes the libraries `Cache` and `LongestPole` and it
+  -- adds `Mathlib/Tactic`.
+  let libs := ← match args.flag? "lib" with
+              | some lib => return #[lib.as! String]
+              | none => getLeanLibs
+  let mut updates := 0
+  for d in libs.reverse do  -- reverse to create `Mathlib/Tactic.lean` before `Mathlib.lean`
+    let fileName := addExtension d "lean"
+    let fileContent ← getAll git d
+    if !(← pathExists fileName) then
+      IO.println s!"Creating '{fileName}'"
+      updates := updates + 1
+      IO.FS.writeFile fileName fileContent
+    else if (← IO.FS.readFile fileName) != fileContent then
+      IO.println s!"Updating '{fileName}'"
+      updates := updates + 1
+      IO.FS.writeFile fileName fileContent
+  if updates == 0 then
+    IO.println "No update necessary"
+  return updates
+
+open Cli in
+/-- Setting up command line options and help text for `lake exe mk_all`. -/
+def mkAll : Cmd := `[Cli|
```

Comment:
Asking for me to learn: why are you quoting `Cli` and not `Cmd`, i.e. why does this line *not* look like this?
`def mkAll : Cmd := `[Cmd|`
That line doesn't work, this one does - if you'd like to tell me, that is welcome. It need not block this review at all.

Reply:
This is completely unsubstantiated speculation: `Cli` makes me think of... well the command-line, while `Cmd` makes me think of a Lean `command`.  Maybe this is the distinction?  So, in my mind, this reads as "I'm defining a `Cmd` (in Lean), by using `Cli` syntax on the command-line".

Reply:
No matter if this is true, it's certainly a useful mnemonic :-)

Reply:
I believe the `Cli` here refers to the [`lean4-cli`](https://github.com/leanprover/lean4-cli) package.

Reply:
:bulb: Now it makes sense! Thanks a lot!